### PR TITLE
Dicodile text example

### PR DESCRIPTION
**Pull request by hndgzkn**
Add a draft example on using `dicodile` on a text image generated with letters PAMI of text_size=5000.

- Creates 10 patches.
- Uses 10 jobs.
- Takes ~45 min. to run `dicodile`.

As the job runs too long, the artifacts can be viewed from https://github.com/hndgzkn/dicodile/actions/runs/624302432

### CODE REVIEW

**Comment by agramfort** [comment]
@hndgzkn can you share an image of the results? thx

**Comment by hndgzkn** [comment]
> @hndgzkn can you share an image of the results? thx

dict_image

text_image

**Comment by agramfort** [comment]
looks great !

>

**Comment by hndgzkn** [comment]
I'd like to document text image generation dependencies: 
- pandoc 
- texlive
- texlive-extra-utils
- texlive-fonts-extra. 

Shall I document it  in **Installation** section of README.rst?

**Comment by agramfort** [comment]
I think this image was produced once for the paper and for illustration
purposes.

I don't think it's really meant to be advertised. I would expect to just
have one image
downloaded for illustration.

>

**Comment by hndgzkn** [comment]
In that case:

- Instead of generating the image, shall I save the image under `data` or `data/images` directory and return this image? 
- Shall I keep the code to generate text?

**Comment by agramfort** [comment]
I would keep the code somewhere to replicate the image and yes store the
image available for download.

basically like we do it for mandrill

>

**Comment by hndgzkn** [comment]
- I created the file `text_4_5000_PAMI.npz` (~18M) and saved it under `dicodile/data/samples` directory. 
- I added a method `load_pami` inside `dicodile/data/images.py` to load PAMI image X and associated D.

I am trying to use gitlfs to push the file, however it is not possible due to [pushing-large-files-to-forks](https://docs.github.com/en/github/managing-large-files/collaboration-with-git-large-file-storage#pushing-large-files-to-forks)

> I would keep the code somewhere to replicate the image and yes store the image available for download. basically like we do it for mandrill
> […](#)

**Comment by agramfort** [comment]
@hndgzkn are you suggesting to add the file to the main github repo? I would not add any big binary file to the source code repo.

I would put the file somewhere on the web eg in 
https://figshare.com/
https://zenodo.org/

File: benchmarks/dicodile_text.py [hunk 1]
```diff
@@ -1,24 +1,22 @@
...
-from dicodile.utils import check_random_state
-
 # Import for online DL
 import spams
 from sklearn.feature_extraction.image import extract_patches_2d
 
 # Import for CDL
 from dicodile import dicodile
-
+from dicodile.data.home import DATA_HOME
```
Comment (agramfort) [inline benchmarks/dicodile_text.py:15]: I would such constants in a config.py file rather than home.py

wdyt @tomMoral ?

File: dicodile/data/_text.py [hunk 2]
```diff
@@ -71,8 +71,21 @@ def convert_str_to_png(text, margin=12):
...
 def get_centered_padding(shape, expected_shape):
-    """Compute a padding to have an array centered in the expected_shape
+    """Compute a padding to have an array centered in the expected_shape.
+
+    Parameters
+    ----------
+    shape: tuple
+      Original array dimensions.
+    expected_shape: ndarray, tuple
+      Expected array dimensions.
```
Comment (agramfort) [inline dicodile/data/_text.py:81]: insert empty line

File: dicodile/data/_text.py [hunk 3]
```diff
@@ -71,8 +71,21 @@ def convert_str_to_png(text, margin=12):
...
+    Returns
+    -------
+    padding: list
+        padding necessary for original array to have the `expected_shape`.
+
```
Comment (agramfort) [inline dicodile/data/_text.py:86]:
```suggestion
```

File: dicodile/data/images.py [hunk 4]
```diff
@@ -18,6 +19,35 @@ def fetch_mandrill():
     return X.swapaxes(0, 2)
 
 
+def fetch_pami():
```
Comment (agramfort) [inline dicodile/data/images.py:22]:
```suggestion
def fetch_letters_pami():
```

File: dicodile/data/images.py [hunk 5]
```diff
@@ -18,6 +19,35 @@ def fetch_mandrill():
...
+    """Loads text image `X` and dictionary `D` of the images of the
+    letters `P`, `A`, `M`, `I` used to generate `X`.
+
+    Returns
+    -------
+
```
Comment (agramfort) [inline dicodile/data/images.py:28]:
```suggestion
```

File: dicodile/data/tests/test_images.py [hunk 6]
```diff
@@ -0,0 +1,12 @@
+from dicodile.data.images import fetch_mandrill, fetch_pami
+
+
+def test_fetch_mandrill():
+    data = fetch_mandrill()
+    assert(3, 512, 512) == data.shape
```
Comment (agramfort) [inline dicodile/data/tests/test_images.py:6]:
```suggestion
    assert (3, 512, 512) == data.shape
```

File: dicodile/data/tests/test_images.py [hunk 7]
```diff
@@ -0,0 +1,12 @@
...
+
+
+def test_fetch_pami():
+    X, D = fetch_pami()
+    assert(2321, 2004) == X.shape
```
Comment (agramfort) [inline dicodile/data/tests/test_images.py:11]:
```suggestion
    assert (2321, 2004) == X.shape
```

File: dicodile/data/tests/test_images.py [hunk 8]
```diff
@@ -0,0 +1,12 @@
...
+    assert(4, 29, 25) == D.shape
```
Comment (agramfort) [inline dicodile/data/tests/test_images.py:12]:
```suggestion
    assert (4, 29, 25) == D.shape
```

File: examples/plot_text.py [hunk 9]
```diff
@@ -0,0 +1,165 @@
+"""DiCoDiLe on text images
+==============================
```
Comment (agramfort) [inline examples/plot_text.py:2]:
```suggestion
"""
DiCoDiLe on text images
==================
```

File: examples/plot_text.py [hunk 10]
```diff
@@ -0,0 +1,165 @@
...
+
+This example illlustrates pattern recovery on a noisy text image using
```
Comment (agramfort) [inline examples/plot_text.py:4]:
```suggestion
This example illustrates pattern recovery on a noisy text image using
```

File: examples/plot_text.py [hunk 11]
```diff
@@ -0,0 +1,165 @@
...
+plt.imshow(extract_x, cmap='gray')
+
+display_dictionaries(D)
+
+###############################################################################
+# We add a Gaussian white noise with standard deviation σ std (X) and σ = 3
+# to `X`.
+
+std = 3
+rng = check_random_state(None)
```
Comment (agramfort) [inline examples/plot_text.py:62]:
```suggestion
rng = np.random.default_rng(None)
```

File: examples/plot_text.py [hunk 12]
```diff
@@ -0,0 +1,165 @@
...
+# set number of patches
+n_atoms = 10
+# set individual atom (patch) size
+atom_support = np.array(D.shape[-2:])
+
+D_init = init_dictionary(X, n_atoms=n_atoms, atom_support=atom_support,
+                         random_state=rng)
+
+# normalize the atoms
+D_init = prox_d(D_init)
```
Comment (agramfort) [inline examples/plot_text.py:79]: I would not expose prox_d here.

Let's just do it with numpy

D_init /= np.sqrt(np.sum(D_init, axis=..., keepdims=True))

File: dicodile/data/tests/test_images.py [hunk 13]
```diff
@@ -0,0 +1,12 @@
+from dicodile.data.images import fetch_mandrill, fetch_pami
```
Comment (hndgzkn) [inline dicodile/data/tests/test_images.py:1]:
```suggestion
from dicodile.data.images import fetch_mandrill, fetch_letters_pami
```

File: dicodile/data/tests/test_images.py [hunk 14]
```diff
@@ -0,0 +1,12 @@
...
+
+
+def test_fetch_mandrill():
+    data = fetch_mandrill()
+    assert (3, 512, 512) == data.shape
+
+
+def test_fetch_pami():
```
Comment (hndgzkn) [inline dicodile/data/tests/test_images.py:9]:
```suggestion
def test_fetch_letters_pami():
```

File: dicodile/data/tests/test_images.py [hunk 15]
```diff
@@ -0,0 +1,12 @@
...
+    X, D = fetch_pami()
```
Comment (hndgzkn) [inline dicodile/data/tests/test_images.py:10]:
```suggestion
    X, D = fetch_letters_pami()
```

File: examples/plot_text.py [hunk 16]
```diff
@@ -0,0 +1,166 @@
...
+
+###############################################################################
+# We will first load PAMI image generated from a text of **5000**
+# characters drawn uniformly from the **4** letters **P** **A** **M**
+# **I** and 2 whitespaces and assign it to `X`.
+#
+# We will also load the images of the four characters used to generate
+# `X` and assign it to variable `D`.
+
+X_original, D = fetch_pami()
```
Comment (hndgzkn) [inline examples/plot_text.py:29]:
```suggestion
X_original, D = fetch_letters_pami()
```

File: examples/plot_text.py [hunk 17]
```diff
@@ -0,0 +1,166 @@
...
+
+This example illustrates pattern recovery on a noisy text image using
+DiCoDiLe algorithm.
+
+"""  # noqa
+import matplotlib.pyplot as plt
+import numpy as np
+
+from dicodile import dicodile
+from dicodile.data.images import fetch_pami
```
Comment (hndgzkn) [inline examples/plot_text.py:13]:
```suggestion
from dicodile.data.images import fetch_letters_pami
```

File: examples/plot_text.py [hunk 18]
```diff
@@ -0,0 +1,166 @@
...
+DiCoDiLe algorithm.
+
+"""  # noqa
+import matplotlib.pyplot as plt
+import numpy as np
+
+from dicodile import dicodile
+from dicodile.data.images import fetch_letters_pami
+from dicodile.update_d.update_d import tukey_window
+from dicodile.utils import check_random_state
```
Comment (hndgzkn) [inline examples/plot_text.py:15]:
```suggestion
```

File: examples/plot_text.py (hunk 12 above)
Comment (hndgzkn) [reply examples/plot_text.py:79]: Can we write a normalization function in utils which could be applied to all dictionaries of atoms?
In Mandrill example we normalize D_init by  D_init = D_init / D_init.max().

File: dicodile/data/_text.py (hunk 2 above)
Comment (hndgzkn) [reply dicodile/data/_text.py:81]: > insert empty line

```suggestion
      Expected array dimensions.
      
```

File: examples/plot_text.py (hunk 12 above)
Comment (agramfort) [reply examples/plot_text.py:79]: D_init = D_init / D_init.max()

seems weird to me. I'll let @tomMoral comment
I would use prox_d there but however I am not fan of exposing
publicly internal jargon like `prox_d`. Nobody will get what it does from function name

File: examples/plot_text.py (hunk 11 above)
Comment (hndgzkn) [reply examples/plot_text.py]: Following this change, I changed 
`X += std * X.std() * rng.randn(*X.shape)` and `noise = noise_level_ * rng.randn(*D_init.shape)` as
`X += std * X.std() * rng.random(X.shape)` and `noise = noise_level_ * rng.random(D_init.shape)` respectively.

This resulted in a noisy initial image and randomly generated dictionary from it as:

noisy

Then the computation lasted 145 min and the result is as follows:
dict


result

[Here](https://numpy.org/doc/stable/reference/random/generated/numpy.random.RandomState.randn.html#numpy.random.RandomState.randn) I see that `RandomState.rand` creates samples  from the “standard normal” distribution. So I tried with:

`X += std * X.std() * rng.standard_normal(X.shape)` and `noise = noise_level_ * rng.standard_normal(D_init.shape)` and this results in:

last

I wanted to verify with you that this is the way to go.

May be we can update random generation using new Generator instead of  [RandomState](https://numpy.org/doc/stable/reference/random/index.html).

File: benchmarks/dicodile_text.py (hunk 1 above)
Comment (tomMoral) [reply benchmarks/dicodile_text.py:15]: Seems like a nice idea to move everything to `dicodile.config.py`.

Comment (hndgzkn) [reply benchmarks/dicodile_text.py:15]: Shall we do it in another pull request, or in this one? It might affect other examples and benchmarks?

File: examples/plot_text.py (hunk 12 above)
Comment (tomMoral) [reply examples/plot_text.py:79]: The `D_init / D_init.max()` is mainly here for display purposes to have images between `0` and `1` and should only be used in `display_dictionaries`.

For this one, I would say that `init_dicitonary` should properly normalize the initial dictionary (with `prox_d` internally no?

Comment (hndgzkn) [reply examples/plot_text.py:79]: Ok. Yes, actually it does. I hadn't checked it, I just followed the code in `benchmarks/dicodile_text.py`. 

Then I'll remove this part.

File: examples/plot_text.py [hunk 19]
```diff
@@ -0,0 +1,165 @@
...
+                         random_state=rng)
+
+# normalize the atoms
+D_init = prox_d(D_init)
+
+# Add a small noise to extracted patches
+noise_level = .1
+noise_level_ = noise_level * D_init.std(axis=(-1, -2), keepdims=True)
+noise = noise_level_ * rng.randn(*D_init.shape)
+D_init = prox_d(D_init + noise)
```
Comment (hndgzkn) [inline examples/plot_text.py:85]: Shall I also remove this part?

Comment (agramfort) [reply examples/plot_text.py:85]: I don't get why you add noise to D_init. So yes I would remove this line. It's the role of the dicitonary init functio to add noise if it needs to

Comment (hndgzkn) [reply examples/plot_text.py:85]: Ok. I just stick to the example `benchmarks/dicodile_text.py`.

File: examples/plot_text.py (hunk 12 above)
Comment (hndgzkn) [reply examples/plot_text.py]: Done!

File: examples/plot_text.py (hunk 19 above)
Comment (hndgzkn) [reply examples/plot_text.py]: Done!

File: examples/plot_text.py [hunk 20]
```diff
@@ -0,0 +1,157 @@
...
+plt.imshow(zoom_x, cmap='gray')
+
+display_dictionaries(D_init)
+
+###############################################################################
+# Add a small noise to avoid having coefficients that are equal which
+# might complicate distributed optimization.
+
+X_0 = X.copy()
+X_0 += X_0.std() * 1e-8 * rng.standard_normal(X.shape)
```
Comment (hndgzkn) [inline examples/plot_text.py:98]: I doubt about the necessity of these lines as well. We already add noise at line 64. This exists in `benchmarks/dicodile_text.py` example. Should it stay?

Comment (hndgzkn) [reply examples/plot_text.py]: Removed as we agreed last week!

**Review by tomMoral — APPROVED** [review]
this looks good to me. A few nitpicks.

Also, maybe add a `print` for the full size of the image when loading it?

**Comment by hndgzkn** [comment]
> this looks good to me. A few nitpicks.
> 
> Also, maybe add a `print` for the full size of the image when loading it?

Initially I had it, but removed as you can hardly see anything. 

whole_image

If you prefer, I can add it.

**Comment by hndgzkn** [comment]
> I was mentionning adding `print(X_orginal.shape)`, so the user see the scale of the problem.

I print the shape for X after reshaping and also I print the shape of dictionary D and D_init as well thinking that it would be also good to see the expected shape for dicodile. If you prefer X_original, I'll change it.

**Review by tomMoral — APPROVED** [review]
LGTM!

**Comment by tomMoral** [comment]
> I print the shape for X after reshaping and also I print the shape of dictionary D and D_init as well thinking that it would be also good to see the expected shape for dicodile. If you prefer X_original, I'll change it.

that is perfect like this. Merging! thanks @hndgzkn :tada:

**Comment by agramfort** [comment]
🎉
